## 실습 1. Excel 파일 불러오기

해야 할 일
 
pandas를 이용해 Excel 파일을 DataFrame으로 불러옵니다.

먼저 코드부터 작성하기보다 인공지능에게 현재 상황과 원하는 작업을 설명해 봅니다.

In [9]:
import pandas as pd  # 데이터 분석에 사용하는 pandas 라이브러리를 불러옵니다

df = pd.read_excel('../../data/raw/교보문고_온라인_베스트셀러_상품리스트.xlsx')  # 같은 폴더에 있는 엑셀 파일을 읽어 df에 저장합니다

df.head()  # df의 앞부분 5개 행을 보여줍니다

c:\dev\llm-data-analysis-course\.venv\Lib\site-packages\openpyxl\styles\stylesheet.py:237: UserWarning: Workbook contains no default style, apply openpyxl's default
  warn("Workbook contains no default style, apply openpyxl's default")


,순위,상품코드,판매상품 ID,상품명,정가,판매가,할인율,적립율,적립예정포인트,인물,출판사,발행(출시)일자,분야
0,1,9791199489561,S000220308313,"세네카, 오늘을 빼앗기고 있는 당신에게","18,000","16,200",10%,5%,900,세네카,논픽션,20260701.0,인문
1,2,9791198742179,S000221143299,홍정기의 장수근육 혁명,"19,800","17,820",10%,5%,990,홍정기,깸,20260831.0,건강
2,3,9788937460586,S000000620195,싯다르타,"8,000","7,200",10%,5%,400,헤르만 헤세,민음사,20020120.0,소설
3,4,9791124038826,S000221075112,머니 트렌드 2027,"27,500","24,750",10%,5%,"1,370",정태익 외,북모먼트,20260916.0,경제/경영
4,5,9791175488854,S000221248222,흔한남매 23,"16,800","15,120",10%,5%,840,흔한남매,미래엔아이세움,20260916.0,어린이(초등)


In [10]:
print(df.shape)       # (행 개수, 열 개수) 확인
print(df.columns.tolist())  # 컬럼명 목록 확인

(989, 13)
['순위', '상품코드', '판매상품 ID', '상품명', '정가', '판매가', '할인율', '적립율', '적립예정포인트', '인물', '출판사', '발행(출시)일자', '분야']


### 결과 확인

- 총 989개 상품, 13개 컬럼으로 구성된 데이터입니다.
- 순위, 상품코드, 상품명, 정가, 판매가, 할인율 등 온라인 서점 베스트셀러 정보를 담고 있습니다.
- 다음 단계에서는 결측치나 데이터 타입 등을 점검해볼 수 있습니다.

## 실습 2. 데이터의 크기와 컬럼 확인하기
 
Excel 파일을 불러왔으면 바로 전처리를 시작하지 않습니다.
 
먼저 어떤 데이터인지 확인합니다.

In [11]:
print(df.shape)          # 1. 전체 행과 컬럼 개수를 (행, 열) 형태로 보여줍니다

print(df.columns)        # 2. 컬럼 이름 목록을 보여줍니다

print(df.dtypes)         # 3. 각 컬럼의 자료형(숫자형, 문자형 등)을 보여줍니다

print(df.isna().sum())   # 4. 각 컬럼별로 결측치(비어있는 값)의 개수를 세어 보여줍니다

df.head()                # 5. 앞의 5행을 표 형태로 보여줍니다

(989, 13)
Index(['순위', '상품코드', '판매상품 ID', '상품명', '정가', '판매가', '할인율', '적립율', '적립예정포인트',
       '인물', '출판사', '발행(출시)일자', '분야'],
      dtype='str')
순위            int64
상품코드            str
판매상품 ID         str
상품명             str
정가              str
판매가             str
할인율             str
적립율             str
적립예정포인트         str
인물              str
출판사             str
발행(출시)일자    float64
분야              str
dtype: object
순위           0
상품코드         0
판매상품 ID      0
상품명          0
정가           0
판매가          0
할인율          0
적립율          0
적립예정포인트      0
인물           3
출판사          3
발행(출시)일자     3
분야          62
dtype: int64


,순위,상품코드,판매상품 ID,상품명,정가,판매가,할인율,적립율,적립예정포인트,인물,출판사,발행(출시)일자,분야
0,1,9791199489561,S000220308313,"세네카, 오늘을 빼앗기고 있는 당신에게","18,000","16,200",10%,5%,900,세네카,논픽션,20260701.0,인문
1,2,9791198742179,S000221143299,홍정기의 장수근육 혁명,"19,800","17,820",10%,5%,990,홍정기,깸,20260831.0,건강
2,3,9788937460586,S000000620195,싯다르타,"8,000","7,200",10%,5%,400,헤르만 헤세,민음사,20020120.0,소설
3,4,9791124038826,S000221075112,머니 트렌드 2027,"27,500","24,750",10%,5%,"1,370",정태익 외,북모먼트,20260916.0,경제/경영
4,5,9791175488854,S000221248222,흔한남매 23,"16,800","15,120",10%,5%,840,흔한남매,미래엔아이세움,20260916.0,어린이(초등)


In [12]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 989 entries, 0 to 988
Data columns (total 13 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   순위        989 non-null    int64  
 1   상품코드      989 non-null    str    
 2   판매상품 ID   989 non-null    str    
 3   상품명       989 non-null    str    
 4   정가        989 non-null    str    
 5   판매가       989 non-null    str    
 6   할인율       989 non-null    str    
 7   적립율       989 non-null    str    
 8   적립예정포인트   989 non-null    str    
 9   인물        986 non-null    str    
 10  출판사       986 non-null    str    
 11  발행(출시)일자  986 non-null    float64
 12  분야        927 non-null    str    
dtypes: float64(1), int64(1), str(11)
memory usage: 218.3 KB


In [13]:
df.isna().sum().sum()

np.int64(71)

### 데이터 구조 확인 결과

- **크기**: 989행 × 13컬럼
- **자료형**: 대부분 문자형(str)이고, `순위`만 정수형(int64), `발행(출시)일자`는 실수형(float64)입니다.
  - `정가`, `판매가`, `할인율` 등 숫자로 보이는 컬럼도 실제로는 문자형이라 계산 전 변환이 필요합니다.
- **결측치**: `분야` 62개, `출판사`·`발행(출시)일자` 각 3개이며, 나머지 컬럼은 결측치가 없습니다.
- **다음 단계**: 가격 관련 컬럼을 숫자형으로 변환하고, `분야` 결측치 처리 방안을 검토합니다.

## 실습 3. 분석에 필요한 컬럼만 선택하기

원본 데이터의 모든 컬럼이 이후 분석에 필요한 것은 아닙니다.

이번 과정에서는 다음 컬럼을 먼저 선택합니다.

In [14]:
cols = ['순위', '판매상품 ID', '상품명', '판매가', '인물', '출판사', '발행(출시)일자', '분야']  # 선택할 컬럼 이름들을 리스트로 정리합니다

df_books = df[cols]  # df에서 원하는 컬럼들만 골라 새로운 DataFrame df_books를 만듭니다

print(df_books.shape)   # 확인 1: (행 개수, 열 개수)가 (989, 8)인지 확인합니다
print(df_books.columns) # 확인 2: 컬럼 이름이 원하는 8개와 일치하는지 확인합니다

df_books.head()          # 확인 3: 실제 데이터가 잘 담겼는지 앞 5행을 눈으로 확인합니다

(989, 8)
Index(['순위', '판매상품 ID', '상품명', '판매가', '인물', '출판사', '발행(출시)일자', '분야'], dtype='str')


,순위,판매상품 ID,상품명,판매가,인물,출판사,발행(출시)일자,분야
0,1,S000220308313,"세네카, 오늘을 빼앗기고 있는 당신에게","16,200",세네카,논픽션,20260701.0,인문
1,2,S000221143299,홍정기의 장수근육 혁명,"17,820",홍정기,깸,20260831.0,건강
2,3,S000000620195,싯다르타,"7,200",헤르만 헤세,민음사,20020120.0,소설
3,4,S000221075112,머니 트렌드 2027,"24,750",정태익 외,북모먼트,20260916.0,경제/경영
4,5,S000221248222,흔한남매 23,"15,120",흔한남매,미래엔아이세움,20260916.0,어린이(초등)


### df_books 생성 결과

- 전체 데이터(13개 컬럼)에서 필요한 8개 컬럼(순위, 판매상품 ID, 상품명, 판매가, 인물, 출판사, 발행(출시)일자, 분야)만 선택했습니다.
- `df_books`는 989행 × 8열로, 컬럼 선택이 의도대로 잘 되었음을 확인했습니다.
- 앞으로의 분석은 원본 `df` 대신 `df_books`를 기준으로 진행합니다.

## 실습 4. 컬럼 이름을 사용하기 쉽게 정리하기

원본 컬럼 이름은 사람이 읽기에는 좋지만 코드에서 반복해서 사용하기에는 조금 길 수 있습니다.

이번 실습에서는 다음처럼 간단하게 정리할 수 있습니다.

In [15]:
df_books = df_books.rename(columns={
    '판매상품 ID': '판매상품ID',
    '인물': '저자',
    '발행(출시)일자': '발행일'
})  # 지정한 컬럼 이름들을 새 이름으로 바꿉니다

print(df_books.columns)  # 컬럼 이름이 잘 바뀌었는지 확인합니다

Index(['순위', '판매상품ID', '상품명', '판매가', '저자', '출판사', '발행일', '분야'], dtype='str')


In [16]:
df_books.columns

Index(['순위', '판매상품ID', '상품명', '판매가', '저자', '출판사', '발행일', '분야'], dtype='str')

### 컬럼 이름 변경 결과

- `판매상품 ID` → `판매상품ID`, `인물` → `저자`, `발행(출시)일자` → `발행일`로 변경했습니다.
- 변경 후 컬럼: 순위, 판매상품ID, 상품명, 판매가, 저자, 출판사, 발행일, 분야
- 컬럼 개수(8개)는 그대로 유지되며, 이름만 의도대로 바뀐 것을 확인했습니다.

## 실습 5. 결측치 확인하기

결측치는 값이 비어 있는 데이터를 의미합니다.

데이터 분석에서는 결측치를 무조건 삭제하지 않습니다.

먼저 어디에 얼마나 있는지 확인하고, 분석 목적에 맞게 처리 방법을 결정합니다.

In [17]:
df_books.isna().sum()  # 컬럼별로 결측치(비어있는 값) 개수를 세어 보여줍니다

순위         0
판매상품ID     0
상품명        0
판매가        0
저자         3
출판사        3
발행일        3
분야        62
dtype: int64

In [18]:
missing = df_books.isna().sum()  # 컬럼별 결측치 개수를 missing에 저장합니다

missing[missing > 0]  # 그중에서 개수가 0보다 큰(결측치가 있는) 컬럼만 골라 보여줍니다

저자      3
출판사     3
발행일     3
분야     62
dtype: int64

### 결측치 확인 결과

- 결측치가 있는 컬럼은 총 4개입니다: 저자(3), 출판사(3), 발행일(3), 분야(62)
- 순위, 판매상품ID, 상품명, 판매가는 결측치가 없습니다.
- `분야` 컬럼의 결측치가 62개로 가장 많아(전체의 약 6%), 추후 분석 시 별도 처리가 필요합니다.

## 실습 6. 판매가를 숫자로 바꾸기

Excel에서 판매가가 다음과 같이 저장되어 있을 수 있습니다.

In [19]:
# 1단계: 쉼표(,)를 빈 문자열로 바꿔서 제거합니다
df_books['판매가'] = df_books['판매가'].str.replace(',', '')

# 2단계: 문자열을 정수형(int)으로 변환합니다
df_books['판매가'] = df_books['판매가'].astype(int)

# 3단계: 변환 후 자료형을 확인합니다
print(df_books['판매가'].dtype)

df_books.head()  # 실제 값도 눈으로 확인합니다

int64


,순위,판매상품ID,상품명,판매가,저자,출판사,발행일,분야
0,1,S000220308313,"세네카, 오늘을 빼앗기고 있는 당신에게",16200,세네카,논픽션,20260701.0,인문
1,2,S000221143299,홍정기의 장수근육 혁명,17820,홍정기,깸,20260831.0,건강
2,3,S000000620195,싯다르타,7200,헤르만 헤세,민음사,20020120.0,소설
3,4,S000221075112,머니 트렌드 2027,24750,정태익 외,북모먼트,20260916.0,경제/경영
4,5,S000221248222,흔한남매 23,15120,흔한남매,미래엔아이세움,20260916.0,어린이(초등)


### 판매가 숫자형 변환 결과

- `판매가` 컬럼의 쉼표를 제거한 뒤 정수형(int64)으로 변환했습니다.
- 변환 후 dtype이 `int64`로 확인되어, 이제 판매가로 계산(합계, 평균, 비교 등)이 가능합니다.
- `head()`로 확인한 값들도 `16200`, `17820`처럼 쉼표 없이 정상적으로 숫자로 표시됩니다.
---

## 실습 7. 발행일을 날짜형으로 바꾸기

원본 데이터의 발행일은 다음처럼 보일 수 있습니다.

In [21]:
# 1단계: 결측치를 허용하는 정수형(Int64)으로 바꿔서 소수점을 없앱니다
df_books['발행일'] = df_books['발행일'].astype('Int64')

# 2단계: 문자열로 바꾼 뒤, YYYYMMDD 형식에 맞춰 날짜로 변환합니다
# errors='coerce'를 넣으면 형식이 안 맞거나 결측치인 값은 에러 대신 NaT(빈 날짜)로 처리됩니다
df_books['발행일'] = pd.to_datetime(df_books['발행일'].astype(str), format='%Y%m%d', errors='coerce')

# 3단계: 변환 후 자료형과 값을 확인합니다
print(df_books['발행일'].dtype)
df_books['발행일'].head()

datetime64[us]


0   2026-07-01
1   2026-08-31
2   2002-01-20
3   2026-09-16
4   2026-09-16
Name: 발행일, dtype: datetime64[us]

### 발행일 날짜형 변환 결과

- `발행일` 컬럼을 `datetime64[us]` 자료형으로 성공적으로 변환했습니다.
- 결측치였던 3개 값은 에러 없이 `NaT`(빈 날짜)로 처리되었습니다.
- 이제 `발행일`을 기준으로 연도/월별 집계나 기간 필터링 등의 날짜 연산이 가능합니다.
---

## 실습 8. 문자열 앞뒤 공백 정리하기

상품명, 저자, 출판사, 분야는 문자열 데이터입니다.

문자열 앞뒤에 불필요한 공백이 있으면 같은 값도 서로 다른 데이터처럼 처리될 수 있습니다.

예를 들어 다음 두 값은 사람이 보기에는 같지만 문자열로는 다릅니다.

In [22]:
cols_to_strip = ['상품명', '저자', '출판사', '분야']  # 공백을 제거할 컬럼 이름들을 리스트로 정리합니다

for col in cols_to_strip:  # 리스트에 담긴 컬럼 이름을 하나씩 꺼내면서 반복합니다
    df_books[col] = df_books[col].str.strip()  # 해당 컬럼의 앞뒤 공백을 제거합니다

df_books[cols_to_strip].head()  # 처리 후 4개 컬럼의 앞 5행을 확인합니다

,상품명,저자,출판사,분야
0,"세네카, 오늘을 빼앗기고 있는 당신에게",세네카,논픽션,인문
1,홍정기의 장수근육 혁명,홍정기,깸,건강
2,싯다르타,헤르만 헤세,민음사,소설
3,머니 트렌드 2027,정태익 외,북모먼트,경제/경영
4,흔한남매 23,흔한남매,미래엔아이세움,어린이(초등)


### 문자열 공백 제거 결과

- `상품명`, `저자`, `출판사`, `분야` 4개 컬럼에 반복문으로 `str.strip()`을 적용해 앞뒤 공백을 제거했습니다.
- `head()`로 확인한 결과 값들이 정상적으로 표시되며, 처리가 잘 적용되었습니다.
- 이제 이 컬럼들로 그룹화나 값 비교를 할 때 공백 때문에 같은 값이 다르게 인식되는 문제를 방지할 수 있습니다.
---

## 실습 9. 중복 데이터 확인하기

데이터에서 중복을 발견했다고 해서 바로 삭제하면 안 됩니다.

먼저 무엇을 기준으로 중복인지 생각해야 합니다.

예를 들어 같은 상품명이 여러 번 등장할 수 있습니다.

하지만 같은 제목이라도 다른 판본, 다른 상품, 개정판 등이 있을 수 있습니다.

따라서 이번 데이터에서는 상품명만 보고 중복 행을 바로 삭제하지 않습니다.

우선 상품을 구분하는 판매상품ID를 확인합니다.

In [23]:
# 1. 전체 행이 완전히 같은 중복 행 개수
print(df_books.duplicated().sum())

# 2. 판매상품ID가 중복된 행 개수
print(df_books.duplicated(subset='판매상품ID').sum())

# 3. 상품명이 중복된 행 개수
print(df_books.duplicated(subset='상품명').sum())

0
0
37


### 중복 데이터 확인 결과

- 전체 행이 완전히 같은 중복: 0건
- 판매상품ID 기준 중복: 0건 → 상품을 구분하는 고유 식별자로 신뢰할 수 있습니다.
- 상품명 기준 중복: 37건 → 같은 제목이 여러 번 등장하지만, 다른 판본·개정판일 수 있으므로 바로 삭제하지 않습니다.

## 실습 10. 전처리 결과 검증하기

지금까지 여러 전처리를 수행했습니다.

마지막에는 반드시 데이터가 의도한 형태가 되었는지 다시 확인합니다.

In [24]:
print("shape:", df_books.shape)
print()

print("columns:", df_books.columns.tolist())
print()

print("dtypes:")
print(df_books.dtypes)
print()

print("isna().sum():")
print(df_books.isna().sum())
print()

print("판매가 앞 5개 값:")
print(df_books['판매가'].head())
print()

print("발행일 앞 5개 값:")
print(df_books['발행일'].head())
print()

print("판매상품ID 중복 개수:", df_books.duplicated(subset='판매상품ID').sum())
print()

df_books.head()  # 전체 데이터 앞 5행을 표로 확인합니다

shape: (989, 8)

columns: ['순위', '판매상품ID', '상품명', '판매가', '저자', '출판사', '발행일', '분야']

dtypes:
순위                 int64
판매상품ID               str
상품명                  str
판매가                int64
저자                   str
출판사                  str
발행일       datetime64[us]
분야                   str
dtype: object

isna().sum():
순위         0
판매상품ID     0
상품명        0
판매가        0
저자         3
출판사        3
발행일        3
분야        62
dtype: int64

판매가 앞 5개 값:
0    16200
1    17820
2     7200
3    24750
4    15120
Name: 판매가, dtype: int64

발행일 앞 5개 값:
0   2026-07-01
1   2026-08-31
2   2002-01-20
3   2026-09-16
4   2026-09-16
Name: 발행일, dtype: datetime64[us]

판매상품ID 중복 개수: 0



,순위,판매상품ID,상품명,판매가,저자,출판사,발행일,분야
0,1,S000220308313,"세네카, 오늘을 빼앗기고 있는 당신에게",16200,세네카,논픽션,2026-07-01,인문
1,2,S000221143299,홍정기의 장수근육 혁명,17820,홍정기,깸,2026-08-31,건강
2,3,S000000620195,싯다르타,7200,헤르만 헤세,민음사,2002-01-20,소설
3,4,S000221075112,머니 트렌드 2027,24750,정태익 외,북모먼트,2026-09-16,경제/경영
4,5,S000221248222,흔한남매 23,15120,흔한남매,미래엔아이세움,2026-09-16,어린이(초등)


### 최종 검증 결과

- 크기: 989행 × 8열, 컬럼 8개 모두 의도한 이름으로 정리됨
- 자료형: 판매가(int64), 발행일(datetime64)로 정상 변환 완료
- 결측치: 저자·출판사·발행일 각 3건, 분야 62건 남아있음(삭제하지 않고 그대로 유지)
- 판매상품ID 중복: 0건으로 고유 식별자 역할 확인
- head() 확인 결과 값들이 전처리 의도대로 잘 반영됨

전처리는 완료됐고, 남은 결측치(특히 분야 62건)는 이후 분석 단계에서 별도로 다루면 됩니다.

## 실습 11. 전처리 데이터 저장하기

다음 Chapter에서는 이 데이터를 다시 사용합니다.

따라서 전처리한 DataFrame을 CSV 파일로 저장합니다.

추천 파일명은 다음과 같습니다.

In [25]:
df_books.to_csv('book_bestseller_clean.csv', index=False, encoding='utf-8-sig')  # 인덱스 없이, 한글이 깨지지 않는 인코딩으로 CSV 파일로 저장합니다

### 전처리 데이터 저장 결과

- 전처리를 마친 `df_books`를 `book_bestseller_clean.csv` 파일로 저장했습니다.
- 인덱스는 저장하지 않았고(`index=False`), `utf-8-sig` 인코딩을 사용해 Windows·엑셀에서도 한글이 깨지지 않습니다.
- 다음 챕터에서는 이 CSV 파일을 불러와서 이어서 분석을 진행합니다.

## 12. 인공지능을 이용해 결과 Markdown 작성하기

실습을 완료하면 Notebook에 이번 작업을 설명하는 Markdown을 작성합니다.

여기서도 인공지능을 사용할 수 있습니다.

단, 인공지능이 실행 결과를 추측하게 하면 안 됩니다.

Notebook에서 실제로 확인한 숫자를 프롬프트에 넣어야 합니다.

### 정리

**1. 무엇을 확인했는지**
원본 데이터는 (989, 13) 크기였고, 이 중 필요한 8개 컬럼만 선택해 `df_books`를 만들었습니다. 결측치를 확인한 결과 저자 3개, 분야 62개가 비어 있었고, 판매상품ID 기준 중복 행은 0개였습니다.

**2. 어떤 전처리를 했는지**
컬럼 이름을 이해하기 쉽게 변경하고, 판매가는 쉼표를 제거한 뒤 int64로, 발행일은 YYYYMMDD 문자열을 datetime64로 변환했습니다. 또한 상품명·저자·출판사·분야 컬럼의 앞뒤 공백을 정리했습니다.

**3. 다음 분석을 위해 어떤 상태가 되었는지**
판매가로 계산이 가능하고, 발행일로 날짜 기반 분석(연도별, 기간별 등)이 가능한 상태가 되었습니다. 저자와 분야에는 아직 결측치가 남아 있어 이후 분석에서 이 부분을 어떻게 다룰지 고려해야 합니다. 전처리된 데이터는 `book_bestseller_clean.csv`로 저장해 다음 챕터에서 이어서 사용할 수 있습니다.